x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java

```diff
@@ -182,12 +182,33 @@ protected void shouldSkipTest(String testName) throws IOException {
 
     protected static void checkCapabilities(RestClient client, TestFeatureService testFeatureService, String testName, CsvTestCase testCase)
         throws IOException {
-        if (testCase.requiredCapabilities.isEmpty()) {
+        if (hasCapabilities(client, testCase.requiredCapabilities)) {
             return;
         }
+
+        var features = Stream.concat(
+            new EsqlFeatures().getFeatures().stream(),
+            new EsqlFeatures().getHistoricalFeatures().keySet().stream()
+        ).map(NodeFeature::id).collect(Collectors.toSet());
```

Comment:
Since you moved this to a better place, do you think it's valid for this Set to be built only once per IT test class instance? Looking at `checkCapabilities` usage I _think_ this Set is called and built for every test query, which seems overkill.

Reply:
This code is called exactly as many times as before, from shouldSkipTest which is run for each test. We could move this collection to a class variable, but it seems a refinement out of scope for this PR.